src/wayland/compositor/mod.rs

```diff
@@ -418,7 +420,24 @@ pub fn add_post_commit_hook(surface: &WlSurface, hook: fn(&DisplayHandle, &WlSur
 ///
 /// It'll be invoked when the surface is destroyed (either explicitly by the client or on
 /// client disconnect).
-pub fn add_destruction_hook(surface: &WlSurface, hook: fn(&SurfaceData)) {
+///
+/// D generic is the compositor state, same as used in `CompositorState::new<D>()`
+pub fn add_destruction_hook<D, F>(surface: &WlSurface, hook: F)
+where
+    F: Fn(&mut D, &SurfaceData) + Send + 'static,
+    D: 'static,
+{
+    let user_state_type = surface.data::<SurfaceUserData>().unwrap().user_state_type;
+    assert_eq!(
```

Comment:
I'm wondering if that should be a debug assert, but downcast_mut already tests the typeids so we wouldn't be eliminating the runtime checks.